frontend/Obelisk/OAuth/Frontend.hs

```diff
@@ -0,0 +1,217 @@
+{-# LANGUAGE TemplateHaskell        #-}
+{-# LANGUAGE TypeApplications       #-}
+
+{-| Reflex component for handling `OAuth` oAuthFrontend.
+
+-}
+module Obelisk.OAuth.Frontend
+  ( -- * Basic interface
+    OAuthFrontendConfig (..)
+  , HasOAuthFrontendConfig (..)
+  , OAuthFrontend (..)
+  , HasOAuthFrontend (..)
+  , TokenGetter
+  , makeOAuthFrontendExeCfg
+  , makeOAuthFrontend
+    -- * Needed constraints
+  , OAuthConstraints
+    -- * Implementation details (probably should not be exported here at all - use with care.)
+  , StoreOAuth (..)
+  , retrieveCode
+  , doAuthorize
+  , retrieveToken
+  ) where
+
+import           Control.Lens
+import           Control.Monad                   ((<=<))
+import           Control.Monad.IO.Class          (MonadIO, liftIO)
+import           Data.Dependent.Sum              (DSum ((:=>)))
+import           Data.Functor.Identity           (Identity (..))
+import           Data.Functor.Identity           (Identity)
+import           Data.Functor.Sum                (Sum (..))
+import           Data.Text                       (Text)
+import qualified GHCJS.DOM                       as DOM
+import qualified GHCJS.DOM.Window                as Window
+import           Language.Javascript.JSaddle     (MonadJSM)
+import           Obelisk.Route                   (Encoder, PageName, R)
+import           Reflex
+
+import           Obelisk.OAuth.Config            (OAuthConfig (..),
+                                                  OAuthConfigPublic,
+                                                  OAuthProvider,
+                                                  getOAuthConfigPublic)
+import           Obelisk.OAuth.Error             (OAuthError (..))
+import           Obelisk.OAuth.Frontend.Internal (makeReflexLenses,
+                                                  tagOnPostBuild)
+import           Obelisk.OAuth.Frontend.Storage
+import           Obelisk.OAuth.Request           (authorizationRequestHref)
+import           Obelisk.OAuth.Route             (OAuthRoute (..),
+                                                  OAuthToken (..),
+                                                  RedirectUriParams (..))
+import           Obelisk.OAuth.State             (OAuthState, genOAuthState)
+
+
+data OAuthFrontendConfig t = OAuthFrontendConfig
+  { _oAuthFrontendConfig_authorize :: Event t ()
+    -- ^ Initiate authorization. Initiate this to get a valid access token.
+  , _oAuthFrontendConfig_route     :: Dynamic t (Maybe (R OAuthRoute))
+    -- ^ Get route updates that are relevant to OAuth.
+  }
+
+makeReflexLenses ''OAuthFrontendConfig
+
+
+data OAuthFrontend t = OAuthFrontend
+  { _oAuthFrontend_authorized :: Dynamic t (Maybe (Either OAuthError OAuthToken))
+    -- ^ When a authorized, this `Dynamic` will hold the needed access token.
+    --   Nothing when not yet authorized, else error or actual token.
+  }
+
+makeReflexLenses ''OAuthFrontend
+
+
+type OAuthConstraints t m =
```

Comment:
Let's not have "Constraints" in the name of a constraint synonym.  (See Joel Spolsky's article on why "Hungarian" notation is bad)

Reply:
I don't really see the problem in this case, also what should I name it then?

Reply:
`HasOAuth` or `SupportsOAuth` or something like that would be good.  Just, generally, the name of a thing shouldn't include the type of the thing (or, in this case, the kind of the thing).

Reply:
Thanks - I like SupportsOAuth :-)